src/ConnectedMode/BranchMatcher.cs

```diff
@@ -0,0 +1,113 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2022 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using LibGit2Sharp;
+using SonarQube.Client;
+
+namespace SonarLint.VisualStudio.ConnectedMode
+{
+    public interface IBranchMatcher
+    {
+        /// <summary>
+        /// Calculates the nearest branch to the head branch.
+        /// Returns null if there are no head branches
+        /// </summary>
+        /// <remarks>
+        /// Compares branches by name and calculates distances by the number of different commits
+        /// Defaults to main branch when no match found
+        /// </remarks>
+        Task<string> GetMatchedBranch(string projectKey);
+    }
+
+    internal class BranchMatcher : IBranchMatcher
+    {
+        private readonly ISonarQubeService sonarQubeService;
+        private readonly IRepository gitRepo;
+
+        public BranchMatcher(ISonarQubeService sonarQubeService, IRepository gitRepo)
```

Comment:
Passing the repo in the constructor means we will need to create a new instance of this class every time the solution changes i.e. we will need a factory to create new instances of this class a runtime.

Alternatively, we could pass the repo in to the `GetMatchedBranch` method.